modules/apps/frontend-js/frontend-js-react-web/test/js/hooks/useIsMounted.tsx

```diff
@@ -15,30 +15,37 @@
 import {cleanup, render} from '@testing-library/react';
 import React from 'react';
 
-import useIsMounted from '../../../src/main/resources/META-INF/resources/js/hooks/useIsMounted.es';
+import useIsMounted from '../../../src/main/resources/META-INF/resources/js/hooks/useIsMounted';
 
 const {useEffect} = React;
 
+type ChildProps = {
+	isMounted?: ReturnType<typeof useIsMounted>;
+};
+
 describe('useIsMounted()', () => {
 	afterEach(cleanup);
 
-	/*
-	 * Regression test for LPS-105721.
-	 */
+	// Regression test for LPS-105721.
+
 	it("can be used in a child component's useEffect callback", () => {
 		let mounted = false;
 
-		const Parent = ({children}) => {
+		const Parent = ({
+			children,
+		}: {
+			children: React.ReactElement<ChildProps>;
+		}) => {
 			const isMounted = useIsMounted();
 
 			const child = React.Children.only(children);
 
 			return React.cloneElement(child, {isMounted});
 		};
 
-		const Child = ({isMounted}) => {
+		const Child = ({isMounted}: ChildProps) => {
 			useEffect(() => {
-				mounted = isMounted();
+				mounted = isMounted!();
```

Comment:
Not sure I understand what does this mean 🤔

Reply:
`!` is [a non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator).

At this point, TS doesn't now that `isMounted` is defined, because it is an optional prop. And it's an "optional prop" only because, unfortunately, `React.Children.only` + `React.cloneElement` are too low level for us to be able to encode the prop types without a cast. Maybe there's a tricky way, but it would probably be pretty verbose.

So, the cast would be ugly. And a non-nullable assertion is ugly too; it's basically me saying to TS, "I know you think that this value could be undefined or null here, and technically you're right from your type system perspective, but I, an organic human being, know better, and I can assure you that there will be a value here that you can call".

The rationale is that this is a test, and if the assertion is invalid, the test will fail anyway, which means we don't have to worry about a mistake going undetected.

Reply:
> The rationale is that this is a test, and if the assertion is invalid, the test will fail anyway, which means we don't have to worry about a mistake going undetected.

So we shouldn't expect this type of assertions in code outside tests?

Reply:
> So we shouldn't expect this type of assertions in code outside tests?

I wouldn't make a lint rule for it, because there are times when it really is the best choice.

In general we should _prefer_ to avoid `any`, `!`, and vague types like `object`, and we're going to have to figure out the best way to provide "strong encouragement" in that direction, but it will require a balance. (It's kind of like code coverage; going for 100% at all costs is probably not going to give you the best return on your investment.)

In the case of something like `!`, it's definitely less of a gaping hole than `any`, which is completely opting out of the type system. It is an assertion, and programming languages and libraries have had assertion mechanisms since the dawn of time for good reason: because even in a statically typed language, there are times where you have to say "I assert that this condition _must_ hold true here, and if I am wrong about that, all bets are off". In the case of `!`, being wrong means you get a runtime error, which is often exactly what you want. Your assumptions are invalidated at that point, and the behavior of the code is undetermined, so aborting is the right thing to do.

Reply:
The cat is already out the bag somewhat in Clay `git grep -w any` returns rather 346 hits, and many of them _are_ type annotations, but I'd like to hold a higher standard inside DXP.

Reply:
I see, thanks for explaining!!

It took me by surprise, and what worries me is how close to valid syntax it is and how easy for someone not knowing what that is to think this is a typo...

Reply:
Well... people are going to have to learn to read TypeScript syntax anyway in order to read/write this stuff, so hopefully this operator will be the smallest of their obstacles... 🤞